File: hs/src/Q800/Q819.hs
module Q800.Q819
(
    mostCommonWord
) where

import qualified Data.Char as Char
import qualified Data.List as List
import qualified Data.HashSet as HashSet

mostCommonWord :: String -> [String] -> String
mostCommonWord paragraph banned = let
                                    ws = words . map Char.toLower . filter (\x -> not (x `elem` ",.;\'!?")) $ paragraph
                                    bans = HashSet.fromList banned
                                    counts = filter (\(x,_)-> not (HashSet.member x bans)) . map (\x -> (x!!0, length x)) . List.group . List.sort $ ws
                                  in
                                    comp counts 0 ""
                                  where
                                    comp :: [(String,Int)] -> Int -> String -> String
                                    comp [] _ ms = ms
                                    comp ((x,xn):xs) mc ms
                                                | xn > mc = comp xs xn x
                                                | otherwise = comp xs mc ms

File: golang/q800/q819.go
package q800

import (
	"regexp"
	"strings"
)

func mostCommonWord(paragraph string, banned []string) string {
	words := splitRegex.Split(strings.ToLower(paragraph), -1)
	bans := map[string]bool{}
	bans[""] = true
	for _, b := range banned {
		bans[b] = true
	}

	counts := map[string]int{}
	for _, w := range words {
		counts[w]++
	}
	max := -1
	r := ""
	for s, c := range counts {
		if c > max {
			if !bans[s] {
				r = s
				max = c
			}
		}
	}
	return r
}

var splitRegex = regexp.MustCompile("[ !?',;.]+")
